src/modules/noit_lua_libnoit_binding.c

```diff
@@ -246,6 +341,169 @@ lua_noit_metric_messages_distributed(lua_State *L) {
+  int rc = noit_metric_parse_tags(name, strlen(name), &stset, &mtset);
+  if (rc < 0) {
+    luaL_error(L, "Error parsing tags for metric %s", name);
+    return 0;
+  }
+  int ref;
+  lua_pushvalue(L, 1);
+  ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  noit_lua_tagset_t l_stset = { .tagset = stset, .lua_name_ref = ref };
+  lua_pushvalue(L, 1);
+  ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  noit_lua_tagset_t l_mtset = { .tagset = mtset, .lua_name_ref = ref };
+  noit_lua_tagset_copy_setup(L, &l_stset);
+  noit_lua_tagset_copy_setup(L, &l_mtset);
+  return 2;
+}
+
+// Eval tag search query against tagset
+// param: ast (userdata)
+// param: tagset (userdata)
+// returns: matches (boolean)
+static int
+lua_noit_tag_search_eval(lua_State *L) {
+  noit_metric_tag_search_ast_t **ast_ud = (noit_metric_tag_search_ast_t **)
+    luaL_checkudata(L, 1, "noit_metric_tag_search_ast_t");
+  noit_metric_tagset_t **set_ud = (noit_metric_tagset_t **)
+    luaL_checkudata(L, 2, "noit_metric_tagset_t");
+  mtev_boolean ok = noit_metric_tag_search_evaluate_against_tags(*ast_ud, *set_ud);
+  lua_pushboolean(L, ok);
+  return 1;
+}
+
+// Eval tag search query against string
+// param: ast (userdata)
+// param: name (string)
+// returns: matches (boolean)
+static int
+lua_noit_tag_search_eval_string(lua_State *L) {
+  noit_metric_tag_search_ast_t **ast_ud = (noit_metric_tag_search_ast_t **)
+    luaL_checkudata(L, 1, "noit_metric_tag_search_ast_t");
+  const char* name = luaL_checkstring(L, 2);
+  noit_metric_tag_t stags[MAX_TAGS], mtags[MAX_TAGS];
+  noit_metric_tagset_t stset = { .tags = stags, .tag_count = MAX_TAGS };
+  noit_metric_tagset_t mtset = { .tags = mtags, .tag_count = MAX_TAGS };
+  int rc = noit_metric_parse_tags(name, strlen(name), &stset, &mtset);
+  if (rc < 0) {
+    luaL_error(L, "Error parsing tagset for metric %s", name);
+    return 0;
+  }
+  mtev_boolean ok = noit_metric_tag_search_evaluate_against_tags(*ast_ud, &mtset);
+  lua_pushboolean(L, ok);
+  return 1;
+}
+
+void
+noit_lua_libnoit_init() {
+  // TODO: Call this once during startup.
+  // I did not find a good place to put this:
+  // - Can't do it from noitd.c: This is a dynamically loaded module.
+  // - Can't do it during module initialization: this is used from the lua_general/lua_web in mtev
+  // - Can't do it during lua state initialization: we have multiple states that get initialized concurrentlya
+  // Also we can't just lock this section with a pthread_mutex: Where should we create the pthread_mutex_t?
+  // Using a spinlock instead, since it does not require initialization
+  ck_spinlock_lock(&noit_lua_libnoit_init_lock);
+  if (account_set == NULL) {
+    mtev_hash_table *tmp = calloc(1, sizeof(*account_set));
+    mtev_hash_init(tmp);
+    ck_pr_store_ptr(&account_set, tmp);
+    metric_director_want_hook_register("metrics_select", hook_metric_subscribe_account, NULL);
+  }
+  ck_spinlock_unlock(&noit_lua_libnoit_init_lock);
+}
```

Comment:
This is the best I could come up with to implement this logic suggested by Theo in chat:
```
pthread_mutex_lock();
if(account_set == NULL) {
  mtec_hash_table *tmptable  = calloc....
  mtec_hash_init(tmptable);
  ck_pr_ptr_store(&account_set, tmptable);
  hook_register...
}
pthead_mutex_unlock();
```
I could not find a good place to initialize the pthread lock, nor where to call this function from during initialization phase. Found the spinlock in CK which can be initialized with a constant.
Not sure if this is a sane solution, but it appears to work 🤞.